tidb/src/main/java/io/tidb/bigdata/tidb/allocator/DynamicRowIDAllocator.java

```diff
@@ -17,52 +17,118 @@
 package io.tidb.bigdata.tidb.allocator;
 
 import io.tidb.bigdata.tidb.ClientSession;
+import io.tidb.bigdata.tidb.meta.TiTableInfo;
 import java.util.Optional;
+import java.util.Random;
 import java.util.concurrent.FutureTask;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
+import org.apache.commons.codec.digest.MurmurHash3;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.tikv.common.meta.TiTimestamp;
 
 public class DynamicRowIDAllocator implements AutoCloseable {
 
   private static final Logger LOG = LoggerFactory.getLogger(DynamicRowIDAllocator.class);
+  private static final ThreadLocal<Random> RANDOM_THREAD_LOCAL =
+      ThreadLocal.withInitial(Random::new);
 
   private final ClientSession session;
   private final String databaseName;
   private final String tableName;
   private final int step;
-  private final long maxShardRowIDBits;
-
+  private final long shardBits;
+  private final TiTimestamp startTimestamp;
+  private final RowIDAllocatorType type;
+  private final boolean isUnsigned;
   private Long start;
   private ThreadPoolExecutor threadPool;
   private FutureTask<Long> futureTask;
   private int index;
+  private long currentShardSeed;
 
   public DynamicRowIDAllocator(
-      ClientSession session, String databaseName, String tableName, int step) {
-    this(session, databaseName, tableName, step, null);
+      ClientSession session,
+      String databaseName,
+      String tableName,
+      int step,
+      TiTimestamp startTimestamp) {
+    this(session, databaseName, tableName, step, null, startTimestamp);
   }
 
   public DynamicRowIDAllocator(
       ClientSession session,
       String databaseName,
       String tableName,
       int step,
-      @Nullable Long start) {
+      @Nullable Long start,
+      TiTimestamp startTimestamp) {
     this.session = session;
     this.databaseName = databaseName;
     this.tableName = tableName;
     this.step = step;
     this.start = start;
-    this.maxShardRowIDBits = session.getTableMust(databaseName, tableName).getMaxShardRowIDBits();
+    this.startTimestamp = startTimestamp;
+    this.type = getType();
+    this.shardBits = getShardBits();
+    this.isUnsigned = isUnsigned();
+    initRandomGenerator();
+  }
+
+  private void initRandomGenerator() {
+    long version = startTimestamp.getVersion();
+    Random random = RANDOM_THREAD_LOCAL.get();
+    random.setSeed(version);
+  }
+
+  private boolean isUnsigned() {
+    TiTableInfo tableInfo = session.getTableMust(databaseName, tableName);
+    switch (type) {
+      case AUTO_INC:
+        return tableInfo.isAutoIncColUnsigned();
```

Comment:
Comparing with other two enum and method, I recommend naming it `AUTO_INCREMENT`.